File: src/emit.c
/******************************************************************************
 *
 *  File Name........: emit.c
 *
 *  Description......:
 *
 *  Created by vin on 03/08/10
 *
 *  Revision History.:
 *
 *****************************************************************************/

#include <stdio.h>
#include <string.h>
#include "emit.h"
#include "ice9.h"

static int dataOffset=1;
static FILE *TMF=NULL;
static int outLine=0;

char *opCodeTab[] = {
    "HALT", "IN", "INB", "OUT", "OUTB", "OUTNL", "ADD", "SUB", "MUL", "DIV", "????",
    /* RR opcodes */
    "LD", "ST", "????",		/* RM opcodes */
    "LDA", "LDC", "JLT", "JLE", "JGT", "JGE", "JEQ", "JNE", "????"
    /* RA opcodes */
};

void emitFile(FILE *f)
{
  TMF = f;
}

int emitSkip(int insts) {
  // use insts = to get current line number
  int r = outLine;
  outLine += insts;
  return r;
}

int emitBackPatch(int line, Inst in, int a, int b, int c, char *note) 
{
  int lineno;

  switch (in) {
  case RRLim:
  case RMLim:
  case RALim:
    CompilerError(-1, "invalid inst opcode %d\n", (int)in);
    return -1;
  default:
    break;
  }

  if (line < 0)
    lineno = outLine++;
  else
    lineno = line;

  fprintf(TMF, "%4d: %7s ", lineno, opCodeTab[in]);
  if (in < RRLim)
    fprintf(TMF, "%d,%d,%d", a, b, c);
  else
    fprintf(TMF, "%d,%d(%d)", a, b, c);

  if (note)
    fprintf(TMF, "\t%s\n", note);
  else
    fputc('\n', TMF);
  return lineno;
}

int emit(Inst in, int a, int b, int c, char *note) 
{
  return emitBackPatch(-1, in, a, b, c, note);
}

int emitComment(char *comment)
{
  fprintf(TMF, "* %s\n", comment);
  return 0;
}

int emitData(int d, char *note)
{
  fprintf(TMF, ".DATA %4d\t%s\n", d, note);
  return dataOffset++;
}

int emitSData(char *s, char *note)
{
  int r = dataOffset;
  dataOffset += strlen(s);
  fprintf(TMF, ".SDATA \"%s\"\t%s\n", s, note);
  return r;
}

int emitDataOffset() {
  return dataOffset;
}
/*........................ end of emit.c ....................................*/

File: python/src/emit.py
#!/usr/bin/python
"""
codegen for ice9

VW Freeh copyright 2010

CSC 512
"""

#########
# Library imports
#########
import sys

#########
# Local imports
#########

#########
# Globals
#########
OutFile = sys.stdout
Comments = True
PC = 0
DataOffset = 1                  # leave room for code size

Instructions = {
    "HALT": "RR",
    "IN": "RR",
    "INB": "RR",
    "OUT": "RR",
    "OUTC": "RR",
    "OUTNL": "RR",
    "ADD": "RR",
    "SUB": "RR",
    "MUL": "RR",
    "DIV": "RR",
    "LD": "RM",
    "ST": "RM",
    "LDA": "RA",
    "LDC": "RA",
    "JLT": "RA",
    "JLE": "RA",
    "JGT": "RA",
    "JGE": "RA",
    "JEQ": "RA",
    "JNE": "RA"
    }

def emitFile(out, no_comments=False):
    global OutFile, Comments
    OutFile = out
    if no_comments:
        Comments = False

def emitSkip(cnt):
    global PC
    r = PC
    PC += cnt
    return r

def emitBackPatch(pc, inst, r, s, t, comment):
    global PC
    try:
        kind = Instructions[inst]
    except KeyError:
        raise CodeGenError(None, 'invalid instruction "%s"', inst)

    if pc < 0:
        pc = PC
        PC += 1
    OutFile.write("%4d:%7s " % (pc, inst))
    if kind == "RR":
        OutFile.write("%d,%d,%d" % (r,s,t))
    else:
        OutFile.write("%d,%d(%d)" % (r,s,t))
    if Comments and comment:
        OutFile.write("\t" + comment)
    OutFile.write("\n")
    return pc

def emit(inst, r, s, t, comment):
    return emitBackPatch(-1, inst, r, s, t, comment)

def emitComment(comment):
    if Comments and comment:
        OutFile.write("* %s\n" % (comment,))

def emitData(val, comment):
    global DataOffset
    OutFile.write(".DATA %4d" % (val,))
    if Comments and comment:
        OutFile.write("\t" + comment)
    OutFile.write("\n")
    r = DataOffset
    DataOffset += 1
    return r

def emitSData(val, comment):
    global DataOffset
    OutFile.write('.SDATA "%s"' % (val,))
    if Comments and comment:
        OutFile.write("\t" + comment)
    OutFile.write("\n")
    r = DataOffset
    DataOffset += len(val)
    return r

def emitDataOffset():
    return DataOffset


